fdbserver/DataDistributionQueue.actor.cpp

```diff
@@ -1807,20 +1958,171 @@ ACTOR Future<bool> rebalanceReadLoad(DDQueueData* self,
+					traceEvent->detail("MultiplyReplicasKeys", metrics.keys.get());
+					self->output.send(RelocateShard(metrics.keys.get(),
+					                                priority,
+					                                RelocateReason::REBALANCE_READ,
+					                                RelocateOperation::MULTIPLY_REPLICAS));
+				} else if (relocateOperation == RelocateOperation::REDUCE_REPLICAS) {
+					traceEvent->detail("ReduceReplicasKeys", metrics.keys.get());
+					self->output.send(RelocateShard(metrics.keys.get(),
+					                                priority,
+					                                RelocateReason::REBALANCE_READ,
+					                                RelocateOperation::REDUCE_REPLICAS));
+				} else {
+					self->output.send(RelocateShard(metrics.keys.get(), priority, RelocateReason::REBALANCE_READ));
+				}
+				self->updateLastAsSource(sourceTeam->getServerIDs());
+				return true;
+			}
+		}
+		traceEvent->detail("SkipReason", "ShardNotPresent");
+		return false;
+	}
+}
+
+ACTOR Future<GetTopKMetricsReply> GetTopOrBottomKMetrics(DDQueueData* self,
+                                                         std::vector<KeyRange> shards,
+                                                         bool isTopK,
+                                                         double srcLoad,
+                                                         double destLoad) {
+	int K = std::min(int(0.1 * shards.size()), SERVER_KNOBS->READ_REBALANCE_SHARD_TOPK);
+	if (isTopK) {
+		//  TODO@ZZX: uncomment
+		GetTopKMetricsRequest req(
+		    // shards, topK, (srcLoad - destLoad) * SERVER_KNOBS->READ_REBALANCE_MAX_SHARD_FRAC, srcLoad /
+		    // shards.size());
+		    shards,
+		    K,
+		    srcLoad);
+		req.comparator = [](const StorageMetrics& a, const StorageMetrics& b) {
+			return a.bytesReadPerKSecond / std::max(a.bytes * 1.0, 1.0) >
+			       b.bytesReadPerKSecond / std::max(b.bytes * 1.0, 1.0);
+		};
+		GetTopKMetricsReply reply = wait(brokenPromiseToNever(self->getTopKMetrics.getReply(req)));
+		return reply;
+	} else { // get BottomK shards
+		GetTopKMetricsRequest req(shards, K, srcLoad / shards.size());
+		req.comparator = [](const StorageMetrics& a, const StorageMetrics& b) {
+			return a.bytesReadPerKSecond / std::max(a.bytes * 1.0, 1.0) <
+			       b.bytesReadPerKSecond / std::max(b.bytes * 1.0, 1.0);
+		};
+
+		GetTopKMetricsReply reply = wait(brokenPromiseToNever(self->getTopKMetrics.getReply(req)));
+		return reply;
+	}
+}
+
+// TODO@ZZX: reduce replicas case
+RelocateOperation selectRelocateOperation(DDQueueData* self,
+                                          TraceEvent* traceEvent,
+                                          DataMovementReason reason,
+                                          double minReadLoad,
+                                          double maxReadLoad,
+                                          double srcLoad,
+                                          StorageMetrics& metrics,
```

Comment:
It is not clear that 'metrics' is for hot shards and `bottomKreply` is for cold shards. Best to give them similar names.

Reply:
Resolve the change, only use reply as before which stands for topK.